File: lua/react/react.lua
local map = function(list, func)
    local result = {}
    for i,elem in ipairs(list) do
        result[i] = func(elem)
    end
    return result
end

local foreach = function(list, func, ...)
    for _, elem in ipairs(list) do
        elem[func](...)
    end
end

------------------------------------------------------------
-- parent of InputCell and ComputeCell
--
local Cell = function()
    local value
    local listeners = {}
    return {
        store_value = function(val) value = val end,

        get_value = function() return value end,

        add_listener = function(listener)
            listeners[#listeners+1] = listener
        end,

        recompute_listeners = function()
            foreach(listeners, "recompute")
        end,

        fire_listener_callbacks = function()
            foreach(listeners, "fire_callbacks")
        end,
    }
end

------------------------------------------------------------
local InputCell = function(val)
    local self = Cell()
    self.store_value(val)

    self.set_value = function(val)
        self.store_value(val)
        self.recompute_listeners()
        self.fire_listener_callbacks()
    end

    return self
end

------------------------------------------------------------
local ComputeCell = function(...)
    local self = Cell()
    local previous
    local inputs = {...}
    local formula = table.remove(inputs)
    local callbacks = {}

    local compute = function()
        local values = map(inputs, function(input)
            return input.get_value()
        end)
        self.store_value(formula(table.unpack(values)))
    end

    -- initialize
    foreach(inputs, "add_listener", self)
    compute()
    previous = self.get_value()

    -- instance methods
    self.recompute = function()
        compute()
        self.recompute_listeners()
    end

    self.fire_callbacks = function()
        local val = self.get_value()
        if previous ~= val then
            previous = val
            for _,callback in ipairs(callbacks) do
                callback(val)
            end
            self.fire_listener_callbacks()
        end
    end

    self.watch = function(callback)
        callbacks[#callbacks+1] = callback
    end

    self.unwatch = function(callback)
        for i, c in ipairs(callbacks) do
            if callback == c then
                table.remove(callbacks, i)
                break
            end
        end
    end

    return self
end

------------------------------------------------------------

return {
    Reactor = function()
        return {
            InputCell = InputCell,
            ComputeCell = ComputeCell,
        }
    end,
}

File: python/react/react.py
class Cell(object):
    def __init__(self, value=None):
        self._value = value
        self._listeners = []

    @property
    def value(self):
        return self._value

    @value.setter
    def value(self, val):
        self._value = val

    def add_listener(self, cell):
        self._listeners.append(cell)

    def recompute_listeners(self):
        for listener in self._listeners:
            listener.recompute()

    def fire_listener_callbacks(self):
        for listener in self._listeners:
            listener.fire_callbacks()


class InputCell(Cell):
    @Cell.value.setter
    def value(self, val):
        self._value = val
        self.recompute_listeners()
        self.fire_listener_callbacks()


class ComputeCell(Cell):
    def __init__(self, inputs, compute_function):
        super().__init__()
        self.inputs = inputs
        for cell in inputs:
            cell.add_listener(self)
        self.formula = compute_function
        self.callbacks = []
        self.compute()
        self.previous = self.value

    def compute(self):
        values = [cell.value for cell in self.inputs]
        self.value = self.formula(values)

    def recompute(self):
        self.compute()
        self.recompute_listeners()

    def fire_callbacks(self):
        val = self.value
        if self.previous != val:
            self.previous = val
            for cb in self.callbacks:
                cb(val)
            self.fire_listener_callbacks()

    def add_callback(self, callback):
        ''' TODO
        should we allow the same callback multiple times?
        '''
        self.callbacks.append(callback)

    def remove_callback(self, callback):
        ''' TODO
        "Guard against incorrect implementations which store
        their callbacks in an array."
        What should the implementation be?
        '''
        while callback in self.callbacks:
            self.callbacks.remove(callback)
